keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+	
+	// ActionName specifies the name of the HookAction that defines the action to take for this hook. 
+	// The HookAction resource type is proposed in https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190120-execution-hook-design.md#implementation-history.
+	ActionName string `json:"actionName,omitempty"`
+}
+
+// DeletionPolicy defines the policy on what happens to an ApplicationSnapshotContent resource and the
+// local application snapshot it represents when the ApplicationSnapshot resource it is bound to is deleted.
+type DeletionPolicy string
+
+// Set of DeletionPolicy values.
+const (
+	DeletionPolicyDelete DeletionPolicy = "Delete"
+	DeletionPolicyRetain DeletionPolicy = "Retain"
+)
+```
+
+`ApplicationSnapshotContent` is a cluster-scoped CRD that describes the details about a local application snapshot. An `ApplicationSnapshotContent` object is created as the result of successfully processing an `ApplicationSnapshot`. The relationship between an `ApplicationSnapshot` object and an `ApplicationSnapshotContent` object is like the relationship between a `VolumeSnapshot` object and a `VolumeSnapshotContent` object. Below is the definition of `ApplicationSnapshotContent`.
+
+```go
+type ApplicationSnapshotContentSpec struct {
+	// SnapshotClassName is the name of the ApplicationSnapshotClass object to use for snapshot-related operations.
+	SnapshotClassName string `json:"snapshotClassName"`
+
+	// SnapshotRef is an reference to the ApplicationSnapshot resource this ApplicationSnapshotContent resource is
+	// bound to. When specified, this creates a bi-directional binding between an ApplicationSnapshot resource and
+	// this ApplicationSnapshotContent resource.
+	// +optional
+	SnapshotRef *corev1.ObjectReference `json:"snapshotRef,omitempty"`
+
+	// VolumeSnapshotRefs is mapping from PVC names to references to the corresponding VolumeSnapshot objects created
+	// for taking snapshots of individual persistent volumes mounted by the application when the snapshot source is
+	// a StatefulApplication.
+	// +optional
+	VolumeSnapshotRefs map[string]corev1.ObjectReference `json:"volumeSnapshotRefs,omitempty"`
+
+	// Tags are user-specified free texts to attach to the snapshot. This combines user-specified tags in the
+	// ApplicationSnapshot and tags returned by pre-hooks.
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
+
+	// MetadataHandle is the handle to persisted application metadata. It contains information needed to read the
+	// metadata back.
+	// +optional
+	MetadataHandle *MetadataHandle `json:"metadataHandle,omitempty"`
+
+	// DeletionPolicy is the policy on what happens to an ApplicationSnapshotContent resource and the
+	// local application snapshot it represents when the ApplicationSnapshot resource it is bound to
+	// is deleted.
+	// DeletionPolicy set here takes precedence over the one (if set) in ApplicationSnapshotClass, if set.
+	// +optional
```

Comment:
To avoid confusion, we have changed all DeletionPolicy fields to Required in the volume snapshot beta APIs so user has to decide exactly what he/she wants.